compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12647,6 +12659,195 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
    return node;
    }
 
+
+/**
+ * \brief
+ * Remove add/subtract operations under comparison nodes to save an
+ * add/sub instruction during codegen.
+ *
+ * \details
+ *
+ * e.g. This following is a before-after comparison.
+ *
+ * \verbatim
+ * Before:
+ * ificmple --> block_17 BBStart at n13n ()
+ *   iadd (X>=0 cannotOverflow )
+ *     ==>iRegLoad (cannotOverflow...)
+ *     iconst -1
+ *   ==>iconst 0
+ *
+ *
+ * After:
+ * ificmple --> block_17 BBStart at n13n ()
+ *   ==>iRegLoad (cannotOverflow...)
+ *   iconst 1
+ *
+ * \endverbatim
+ *
+ *
+ * The iadd node should be marked as 'cannotOverflow' to ensure functional correctness.
+ *
+ * \param node          the comparison node being simpilfied
+ * \param s             the simplifier object
+ * \param isInt         true if the integral types is less than or euqal to 32-bit wide;
+ *                      false if it's 64-bit.
+*/
+TR::Node* removeArithmeticsUnderIntegralCompare(TR::Node* node,
+                                                TR::Simplifier * s,
+                                                bool isInt)
+   {
+   TR::Node* firstChild     = node->getFirstChild();
+   TR::Node* secondChild    = node->getSecondChild();
+   TR::ILOpCodes op         = firstChild->getOpCodeValue();
+
+   bool isAdd = (op == TR::iadd || op == TR::ladd || op == TR::iuadd || op == TR::luadd ||
+                 op == TR::sadd || op == TR::badd || op == TR::buadd);
+
+   bool isSub = (op == TR::isub || op == TR::lsub || op == TR::iusub || op == TR::lusub ||
+                 op == TR::ssub || op == TR::bsub || op == TR::busub);
+
+   if ((isAdd || isSub) &&
+           firstChild->chkCannotOverflow()  &&
+           firstChild->getSecondChild()->getOpCode().isLoadConst() &&
+           secondChild->getOpCode().isLoadConst() &&
+           firstChild->isSingleRef() &&
+           !(!firstChild->isSingleRef() &&
+             firstChild->getFutureUseCount() != firstChild->getReferenceCount() - 1))
+      {
+      bool isUnsigned = firstChild->getOpCode().isUnsigned();
```

Comment:
As Andrew mentioned above, addition and subtraction are independent of signedness, and the unsigned opcodes are deprecated. The relevant thing here would be whether `node->isUnsignedCompare()`.

Unfortunately, `cannotOverflow` tells us that the operation doesn't signed-overflow, and I don't think it has an unsigned version, so we may have to give up on unsigned inequalities.